### DISCUSSION TITLE
Export a table in DuckDB to a Parquet file via hive style partitioning

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#6213] [Q&A] [answered]
Hello everyone,

I was looking DuckDB's online documentation, and at chapter [Parquet](https://duckdb.org/docs/data/parquet.html), I found a very interesting feature allowing to export a table within the database, **directly into the disk** converted to Apache Parquet format. 

`COPY (SELECT * FROM tbl) TO 'result-snappy.parquet' (FORMAT 'parquet')`

This could be very useful for me as I work in a shared/multi-user Datalab where the amount of available RAM is not always enough to put an entire dataset in-memory (for example inside a Data.frame in a R program) before exporting it to Parquet on the disk. So this direct export is really handy. 

I would like to know whether there is any possibility to export a table from DuckDB to Parquet via `Hive style` by specifying keys. As you can see, the above `COPY` command just builds the whole thing inside a single file without any possibility of specifying keys based on which the export is split. On the other hand, using the Arrow package in R, there is a function called `write_dataset` with an argument called **partitioning** that lets us to define keys based on specific columns via **Hive style** so at the end it is possible to have a multi-file dataset.

Is it possible to do the same thing with DuckDB? Just to be clear, I am talking about the export ( = creating Parquet) **from DuckDB to the local file system** and not the import ( = reading Parquet).

Thanks in advance.

**Comment by Mytherin** [comment, marked as answer]
Writing hive-partitioned Parquet files is supported in the latest development release and will be supported in the next release (see #5964).

**Reply by parvizfarnia** [reply]
Thank you very much for your quick answer. This is really an excellent news.

As I am completely a beginner in DuckDB and just to be sure that I understand correctly I explain my situation in a few lines : I work in a shared multi-user Datalab environment where I run my R programs on often big CSV files (>= 10 GB). We have a lot of disk available (hunderds of GB) but **often far less RAM available** given the number of users who are connected and working simultaneously in the same environment. Consequently I try to avoid Data.frames and in general, any in-memory data structure as much as possible. Instead I import my data in databases such as SQLite and often, only the result of my queries would go into RAM as data.frame. By proceeding like this, so far, I have been able to deal with the problem of limited RAM. 

Recently I started working with Parquet and I had to look for a new kind of database because as I understand there is no direct way to export SQLite tables into Parquet. Once again, I insist on database and not in-memory because during my tests I noticed that producing large datasets with partitioning based on keys upon big data.frames (in-memory) can indeed consume tremendous amount of RAM (apart from the data.frame itself) as is it seems that it tries to build some kind of hierarchical structure (Tree) and only then it would flush the data on the disk. That's why I asked for advice on StackOverflow and I was told that DuckDB was the way to go as creating a parquet with the COPY command is almost disk to disk ( = from database to created parquet file) and compared to Data.frames, far less memory on RAM will be consumed in order to create Parquet.

I was looking the link that you provided above and the following paragraph gained my attention: 

> The partitioned write fully materializes, **so it will quickly fill the available memory for large tables**. This is not necessarily a blocker as the buffer manager can offload to disk. This will however mean that enough local disk space is required and for large files the data is:
> 
>     stored in memory
>     offloaded to disk
>     read from disk
>     written to the final partition file
> 
> This is not ideal, but will still be good enough for many use cases.

Therefore, I'm not sure to have understood the concept. Am I going to have problems with RAM or the job is done mainly on the disk at the database level?

Thank you very much

**Reply by Mytherin** [reply]
DuckDB allows you to set the maximum amount of memory it can consume using the `SET memory_limit='XGB'` command. DuckDB will then use memory as required until the memory limit is reached, and offload blocks to disk to ensure it stays under the memory limit. That means that setting the memory limit lower will result in a performance penalty as disk I/O is performed - but memory pressure will be lower.

Note that there are limitations to this. If the memory limit is too low the query might not be able to complete at all as not all blocks can be evicted to disk. There are also certain intermediate structures that are not counted towards this memory limit currently.

**Reply by parvizfarnia** [reply]
Thank you very much for this clarification. I can negotiate to have at least 10GB of RAM and I have at least 300 GB hard disk space exclusively for my own programs. As I said, administrators are quite generous for hard disk space but far less generous for RAM (given the number of users working simultaneously). But if I can use disk space instead of RAM even by rather a degraded performance (I mean longer time to build the Parquet) I think I can live with that, because I know that at least I would be able to get the job done. This is going to be truly a life-saver for me! Just one last question before closing this thread, because you said this new functionality, that is, `writing hive-partitioned Parquet` is going to be available in the next release, is there any date (even approximative) indicating when the functionality would be available in Python and/or R ? Because I will have to call the DuckDB's `COPY` command within R and/or Python programs.

Thanks in advance.

**Reply by Mytherin** [reply]
We are pushing the next release next week (likely tomorrow).

**Reply by parvizfarnia** [reply]
At the moment I have the following version in R

```
> packageVersion("duckdb")
[1] '0.6.2'
>

```

So during next week, I'll check for the new version to install.

Thank you very much for your time and your help. Much appreciated and have a nice week.

Regards.
